include/mcl_3dl/point_cloud_random_samplers/point_cloud_sampler_with_normal.h

```diff
@@ -0,0 +1,211 @@
+    normal_search_range_ = normal_search_range;
+  }
+
+  void setParticleStatistics(const State6DOF& mean, const std::vector<State6DOF>& covariances)
+  {
+    mean_ = mean;
+    Matrix pos_cov(3, 3);
+    for (size_t i = 0; i < 3; ++i)
+    {
+      for (size_t j = 0; j < 3; ++j)
+      {
+        pos_cov(i, j) = std::abs(covariances[i][j]);
+      }
+    }
+    const Eigen::SelfAdjointEigenSolver<Matrix> eigen_solver(pos_cov);
+    eigen_vectors_ = eigen_solver.eigenvectors();
+    eigen_values_ = eigen_solver.eigenvalues();
+  }
+
+  typename pcl::PointCloud<POINT_TYPE>::Ptr sample(
+      const typename pcl::PointCloud<POINT_TYPE>::ConstPtr& pc,
+      const size_t num) const final
+  {
+    const ros::WallTime start_timestamp = ros::WallTime::now();
+
+    typename pcl::PointCloud<POINT_TYPE>::Ptr output(new pcl::PointCloud<POINT_TYPE>);
+    output->header = pc->header;
+
+    if (pc->points.size() == 0)
+    {
+      return output;
+    }
+    if (pc->size() <= num)
+    {
+      *output = *pc;
+      return output;
+    }
+
+    const double eigen_value_ratio = std::sqrt(eigen_values_[2]) / std::sqrt(eigen_values_[1]);
+
+    double max_weight = 1.0;
+    if (eigen_value_ratio < perform_weighting_ratio_)
+    {
+      max_weight = 1.0;
+    }
+    else if (eigen_value_ratio > max_weight_ratio_)
+    {
+      max_weight = max_weight_;
+    }
+    else
+    {
+      const double weight_ratio =
+          (eigen_value_ratio - perform_weighting_ratio_) / (max_weight_ratio_ - perform_weighting_ratio_);
+      max_weight = 1.0 + (max_weight_ - 1.0) * weight_ratio;
+    }
+    const mcl_3dl::Vec3 fpc_global(eigen_vectors_(0, 2), eigen_vectors_(1, 2), eigen_vectors_(2, 2));
+    const mcl_3dl::Vec3 fpc_local = mean_.rot_.inv() * fpc_global;
+    pcl::NormalEstimation<POINT_TYPE, pcl::Normal> ne;
+    pcl::PointCloud<pcl::Normal>::Ptr cloud_normals(new pcl::PointCloud<pcl::Normal>);
+    typename pcl::search::KdTree<POINT_TYPE>::Ptr tree(new pcl::search::KdTree<POINT_TYPE>());
+    ne.setInputCloud(pc);
+    ne.setSearchMethod(tree);
+    ne.setRadiusSearch(normal_search_range_);
+    ne.compute(*cloud_normals);
+
+    const ros::WallTime compute_normal_timestamp = ros::WallTime::now();
+    std::vector<double> cumulative_weight(cloud_normals->points.size(), 0.0);
+    for (size_t i = 0; i < cloud_normals->points.size(); i++)
+    {
+      double weight = 1.0;
+      const auto& normal = cloud_normals->points[i];
+      if (!std::isnan(normal.normal_x) && !std::isnan(normal.normal_y) && !std::isnan(normal.normal_z))
+      {
+        double acos_angle = std::abs(normal.normal_x * fpc_local.x_ +
+                                     normal.normal_y * fpc_local.y_ +
+                                     normal.normal_z * fpc_local.z_);
+        // Avoid that std::acos() returns nan because of calculation errors
+        if (acos_angle > 1.0)
+        {
+          acos_angle = 1.0;
+        }
+        const double angle = std::acos(acos_angle);
```

Comment:
I'd like to avoid trigonometric functions in the loop of the input points as much as possible.

Reply:
I agree, but it takes less than 0.2ms to process this block on my laptop, and the optimization benefits will be small if it is removed.

Reply:
It looks fine as it is not a default option.
Maybe it's better to add short description to algorithm and parameter documents later.